src/client/pythonEnvironments/discovery/locators/services/pipEnvHelper.ts

```diff
@@ -0,0 +1,127 @@
+        if (maxDepth) {
+            depthToSearch = +maxDepth;
+        } else {
+            depthToSearch = 3;
+        }
+    }
+    while (depthToSearch > 0 && searchDir !== path.dirname(searchDir)) {
+        const pipFile = path.join(searchDir, pipFileName);
+        // eslint-disable-next-line no-await-in-loop
+        if (await pathExists(pipFile)) {
+            return pipFile;
+        }
+        searchDir = path.dirname(searchDir);
+        depthToSearch -= 1;
+    }
+    return undefined;
+}
+
+/**
+ * If interpreter path belongs to a pipenv environment which is located inside a project, return associated Pipfile,
+ * otherwise return `undefined`.
+ * @param interpreterPath Absolute path to any python interpreter.
+ */
+async function getPipfileIfLocalPipenvEnvironment(interpreterPath: string): Promise<string | undefined> {
+    // Local pipenv environments are created by setting PIPENV_VENV_IN_PROJECT to 1, which always names the environment
+    // folder '.venv': https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.PIPENV_VENV_IN_PROJECT
+    // This is the layout we wish to verify.
+    // project
+    // |__ Pipfile  <--- check if Pipfile exists here
+    // |__ .venv    <--- check if name of the folder is '.venv'
+    //     |__ Scripts/bin
+    //         |__ python  <--- interpreterPath
+    const venvFolder = path.dirname(path.dirname(interpreterPath));
+    if (path.basename(venvFolder) !== '.venv') {
+        return undefined;
+    }
+    const directoryWhereVenvResides = path.dirname(venvFolder);
+    return getAssociatedPipfile(directoryWhereVenvResides, false);
+}
+
+/**
+ * If interpreter path belongs to a global pipenv environment, return associated Pipfile, otherwise return `undefined`.
+ * @param interpreterPath Absolute path to any python interpreter.
+ */
+async function getPipfileIfGlobalPipenvEnvironment(interpreterPath: string): Promise<string | undefined> {
+    // Global pipenv environments have a .project file with the absolute path to the project
+    // See https://github.com/pypa/pipenv/blob/9299ae1f7353bdd523a1829f3c7cad0ee67c2e3b/CHANGELOG.rst#L754
+    // Also, the name of the directory where Pipfile resides is used as a prefix in the environment folder.
+    // This is the layout we wish to verify.
+    // <Environment folder>
+    // |__ .project  <--- check if .project exists here
+    // |__ Scripts/bin
+    //     |__ python  <--- interpreterPath
+    const dotProjectFile = path.join(path.dirname(path.dirname(interpreterPath)), '.project');
+    if (!(await pathExists(dotProjectFile))) {
+        return undefined;
+    }
+
+    const project = await readFile(dotProjectFile);
+    if (!(await pathExists(project))) {
+        return undefined;
+    }
+
+    // The name of the directory where Pipfile resides is used as a prefix in the environment folder.
+    if (interpreterPath.indexOf(`${path.sep}${path.basename(project)}-`) === -1) {

```

Comment:
It's unclear from the comments and code what the relationship is between `interpreterPath`, `project`, and the `<Environment folder>`.  `interpreterPath` is by definition found under `<Environment folder>`, so this line seems to be checking if `project` is a parent of `<Environment folder>`.  However, that does not seem like the right check.  Is this meant to follow symlinks or something?

Reply:
I'll clarify. I think you're confusing what `project` refers to. `project` is the directory where Pipfile resides. And Pipfile resides outside the `<Environment folder>`.

> interpreterPath is by definition found under <Environment folder>

That's correct

> so this line seems to be checking if project is a parent of <Environment folder>

If the project is say, `project3`. The environment folder is named something like `project3-<hash>` for global environments. So this line is checking if interpreter path string contains a substring `/project3-<hash>`.

Reply:
I have added more comments to elaborate, please have a look again.

Reply:
So the project dir is always an ancestor to the env folder?

```
/x/y/z/project3  # project dir
           |- Pipfile
           |- w
              |- project3-cb61d3e2
                  |- my-env  # env root
                      |- .project
                      |- Scripts/bin
                          |- python  # interpreterPath
```

If that is the case then why check `interpreterPath`?  Isn't checking `envRoot` more direct?  Checking `interpreterPath` implies that it has information that `envRoot` does not, which it does not (If I've understood correctly).

So:

```suggestion
    if (envRoot.indexOf(`${path.sep}${path.basename(project)}-`) === -1) {
```

Or is the env root always named `<project base>-<hash>`?

```
/x/y/z/project3  # project dir
           |- Pipfile
/w/my-workspace
      |- project3-cb61d3e2  # env root
          |- .project
          |- Scripts/bin
              |- python  # interpreterPath
```

If so then why not use this check:

```suggestion
    if (envRoot.match(new RegExp(`.*${path.sep}${path.basename(project)}-[a-fA-F0-9]+$`)) !== null) {
```

Or for a little more readability:

```suggestion
    const pattern = `\\${path.sep}${path.basename(project)}-[a-fA-F0-9]+\b`;
    if (envRoot.match(new RegExp(`^.*${pattern}$`)) !== null) {
```

Or is there some other tight relationship between the project dir and the env root?

Reply:
If we were aiming for reusability with the pattern, I'd suggest adding a lookahead:

```suggestion
    const pattern = `\\${path.sep}${path.basename(project)}-[a-fA-F0-9]+(?=\\${path.sep}|$)`;
    if (envRoot.match(new RegExp(`^.*${pattern}$`)) !== null) {
```

However, we are not. 😄

Reply:
So the project dir is always an ancestor to the env folder?

```
/x/y/z/project3  # project dir
           |- Pipfile
           |- w
              |- project3-cb61d3e2
                  |- my-env  # env root
                      |- .project
                      |- Scripts/bin
                          |- python  # interpreterPath
```

If that is the case then why check `interpreterPath`?  Isn't checking `envRoot` more direct?  Checking `interpreterPath` implies that it has information that `envRoot` does not, which it does not (If I've understood correctly).

So:

```suggestion
    if (envRoot.indexOf(`${path.sep}${path.basename(project)}-`) === -1) {
```

Or is the env root always named `<project base>-<hash>`?

```
/x/y/z/project3  # project dir
           |- Pipfile
/w/my-workspace
      |- project3-cb61d3e2  # env root
          |- .project
          |- Scripts/bin
              |- python  # interpreterPath
```

If so then why not use this check:

```suggestion
    if (envRoot.match(new RegExp(`.*${path.sep}${path.basename(project)}-[a-fA-F0-9]+$`)) !== null) {
```

Or for a little more readability:

```suggestion
    const pattern = `\\${path.sep}${path.basename(project)}-[a-fA-F0-9]+\b`;
    if (envRoot.match(new RegExp(`^.*${pattern}$`)) !== null) {
```

Or is there some other tight relationship between the project dir and the env root?

Reply:
If we were aiming for reusability with the pattern, I'd suggest adding a lookahead:

```suggestion
    const pattern = `\\${path.sep}${path.basename(project)}-[a-fA-F0-9]+(?=\\${path.sep}|$)`;
    if (envRoot.match(new RegExp(`^.*${pattern}$`)) !== null) {
```

However, we are not. 😄

Reply:
Hmm, I think I accidentally deleted my actual comment...

I wanted to clarify.  If I've understood correctly, the directory structure is like this:

```
/x/y/project3  # project dir
       |- Pipfile
/x/z/my-workspace  # workspace root
       |-project3-3f21b0aa  # env root
           |- .project
           |- Scripts/bin
               |- python  # interpreterPath
```

If that is the case, then why not check the env root?  Otherwise it makes it seem like there's something special to `interpreterPath`.

```suggestion
    const pattern = `\\${path.sep}${path.basename(project)}-[a-fA-F0-9]+`;
    if (envRoot.match(new RegExp(`^.*${pattern}$`)) !== null) {
```

Reply:
Right. I have changed the code to compare environment root instead of interpreter path. However I don't think using regex here is necessary, so I haven't included that.